### DISCUSSION TITLE
How to use dagster for real time inference ?

### DISCUSSION THREAD

**Discussion by RBrossard** [dagster-io/dagster#4883] [Q&A]
Hi, I am using Dagster as an orchestrator for all of my training pipeline of ML and really loving it. Thank you for the great work !

I am now thinking about using dagster for real time inference pipeline, but I am not sure about how I should proceed. I have two main concerns:

1) Executing dagster has quite some overhead. For example, a pipeline with a few empty nodes  is executed in 100 ms, which is far too much for real time inference: 
```python
from time import time
from dagster import pipeline, solid, execute_pipeline


@solid
def foo(): pass


@pipeline
def bar():
    for i in range(50):
        foo()


run_config = {"loggers": {"console": {"config": {"log_level": "ERROR"}}}}

t0 = time()
execute_pipeline(bar, run_config=run_config)
print(time()-t0)
```
I am not sure to understand what takes so much time. Is it all the checks that dagster is doing ? In any case, I am happy to use all the features while testing, but once the pipeline is robust and ready for production, I will only be interested into passing the solid_configs and resources. Is there a way to deactivate all of the heavy bunch ?

2) The execution overhead is not a problem if it's only at initialization and I can stream data to the pipeline. I thought about using the dynamic output for that, thinking it would work like a generator. However this seems not to be the case:

```python
from dagster import pipeline, solid, execute_pipeline
from dagster import DynamicOutput, DynamicOutputDefinition, pipeline, solid


@solid(
    output_defs=[DynamicOutputDefinition(int)],
)
def first_step():
    for i in range(3):
        print("first step", i)
        yield DynamicOutput(i, mapping_key=str(i))


@solid
def second_step(value: int) -> int:
    print("second step", value)
    return value


@pipeline
def pipe():
    generator = first_step().map(second_step)


run_config = {"loggers": {"console": {"config": {"log_level": "ERROR"}}}}

execute_pipeline(pipe, run_config=run_config)
```
The previous code prints:
```
first step 0
first step 1
first step 2
second step 0
second step 1
second step 2
```
while I would hope that it behaves like a generator, meaning `second_step 0` printed just after `first_step 0`. In a more advanced example, the first step would yield a stream of data in real time and the rest of the pipeline would have to run as soon as it receives it.

With my current  understanding, I have no other idea for real time inference using dagster. Am I missing something ? What do you think ?

**Comment by sryza** [comment]
Hi @RBrossard - I meant to respond to this earlier.  Regarding the 100 ms latency you're observing, there's no fundamental architectural reason that it should take so long, but we'd likely need to do profiling to speed it up.  I can't promise that's something will be able to get too soon, but would be happy to help out if you wanted to look into it.

Regarding dynamic outputs, what you observed is expected - we don't start downstream steps until a step has fully completed.  We've talked about improving that in the future, but it's not something we have immediate plans for.
